ddtrace/contrib/wsgi/_utils.py

```diff
@@ -0,0 +1,125 @@
+import functools
+import sys
+from typing import TYPE_CHECKING
+
+from ddtrace.internal.compat import PY2
+
+
+if TYPE_CHECKING:
+    from typing import Any
+    from typing import Callable
+    from typing import Dict
+    from typing import Iterable
+
+    from ddtrace import Pin
+    from ddtrace import Span
+    from ddtrace import Tracer
+    from ddtrace.settings import Config
+
+
+if PY2:
+    import exceptions
+
+    generatorExit = exceptions.GeneratorExit
+else:
+    import builtins
+
+    generatorExit = builtins.GeneratorExit
+
+
+import six
+
+from ddtrace.ext import SpanTypes
+
+from .. import trace_utils
+
+
+__metaclass__ = type
+
+
+class DDWSGIMiddlewareBase:
+    """Base WSGI middleware class.
+
+    :param application: The WSGI application to apply the middleware to.
+    :param tracer: Tracer instance to use the middleware with. Defaults to the global tracer.
+    :param int_config: Integration specific configuration object.
+    :param pin: Set tracing metadata on a particular traced connection
+    """
+
+    def __init__(self, application, tracer, int_config, pin):
+        # type: (Iterable, Tracer, Config, Pin) -> None
+        self.app = application
+        self.tracer = tracer
+        self.config = int_config
+        self.pin = pin
+
+    @property
+    def request_span_name(self):
+        # type: () -> str
+        raise NotImplementedError
+
+    @property
+    def application_span_name(self):
+        # type: () -> str
+        raise NotImplementedError
+
+    @property
+    def response_span_name(self):
+        # type: () -> str
+        raise NotImplementedError
+
+    def __call__(self, environ, start_response):
+        # type: (Iterable, Callable) -> None
+        trace_utils.activate_distributed_headers(self.tracer, int_config=self.config, request_headers=environ)
+
+        with self.tracer.trace(
+            self.request_span_name,
+            service=trace_utils.int_service(self.pin, self.config),
+            span_type=SpanTypes.WEB,
+        ) as req_span:
+            # This prevents GeneratorExit exceptions from being propagated to the top-level span.
```

Comment:
aside: I wonder if we should be ignoring GeneratorExit exceptions by default in spans since I think they're always erroneous?